pkg/sentry/fsimpl/erofs/erofs.go

```diff
@@ -0,0 +1,362 @@
+	image *erofs.Image
+}
+
+// Name implements vfs.FilesystemType.Name.
+func (FilesystemType) Name() string {
+	return Name
+}
+
+// Release implements vfs.FilesystemType.Release.
+func (FilesystemType) Release(ctx context.Context) {}
+
+// GetFilesystem implements vfs.FilesystemType.GetFilesystem.
+func (fstype FilesystemType) GetFilesystem(ctx context.Context, vfsObj *vfs.VirtualFilesystem, creds *auth.Credentials, source string, opts vfs.GetFilesystemOptions) (*vfs.Filesystem, *vfs.Dentry, error) {
+	mopts := vfs.GenericParseMountOptions(opts.Data)
+
+	fd, err := getFDFromMountOptionsMap(ctx, mopts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	f := os.NewFile(uintptr(fd), "image file")
+	image, err := erofs.OpenImage(f)
+	if err != nil {
+		f.Close()
+		return nil, nil, err
+	}
+
+	devMinor, err := vfsObj.GetAnonBlockDevMinor()
+	if err != nil {
+		image.Close()
+		return nil, nil, err
+	}
+
+	fs := &filesystem{
+		mopts:    opts.Data,
+		image:    image,
+		devMinor: devMinor,
+	}
+	fs.vfsfs.Init(vfsObj, &fstype, fs)
+
+	root := fs.newDentry(image.Root())
+	// Increase the root's reference count to 2. One reference is returned to
+	// the caller, and the other is held by fs.
+	root.IncRef()
+	fs.root = root
+
+	return &fs.vfsfs, &root.vfsd, nil
+}
+
+func getFDFromMountOptionsMap(ctx context.Context, mopts map[string]string) (int, error) {
+	ifdstr, ok := mopts[moptImageFD]
+	if !ok {
+		ctx.Warningf("erofs.getFDFromMountOptionsMap: image FD must be specified as '%s=<file descriptor>'", moptImageFD)
+		return -1, linuxerr.EINVAL
+	}
+	delete(mopts, moptImageFD)
+
+	ifd, err := strconv.Atoi(ifdstr)
+	if err != nil {
+		ctx.Warningf("erofs.getFDFromMountOptionsMap: invalid image FD: %s=%s", moptImageFD, ifdstr)
+		return -1, linuxerr.EINVAL
+	}
+
+	return ifd, nil
+}
+
+// dentry implements vfs.DentryImpl.
+//
+// +stateify savable
+type dentry struct {
+	vfsd vfs.Dentry
+
+	dentryRefs
+
+	// fs is the owning filesystem.
+	fs *filesystem
+
+	// parent is this dentry's parent directory. If this dentry is
+	// a file system root, parent is nil.
+	parent *dentry
+
+	// name is this dentry's name in its parent. If this dentry is
+	// a file system root, name is the empty string.
+	name string
+
+	// inode is the inode represented by this dentry in image.
+	inode *erofs.Inode
+
+	// dirMu serializes changes to the dentry tree.
+	dirMu sync.RWMutex `state:"nosave"`
+
+	// childMap contains the mappings of child names to dentries if
+	// this dentry represents a directory.
+	// +checklocks:dirMu
+	childMap map[string]*dentry
+
+	// mf implements memmap.File for mappings of this dentry.
+	mf dentryMemmapFile
+
+	locks vfs.FileLocks
+
+	// Inotify watches for this dentry.
+	watches vfs.Watches
+}
+
+// DecRef implements vfs.DentryImpl.DecRef.
+func (d *dentry) DecRef(ctx context.Context) {
+	d.dentryRefs.DecRef(func() {
+		d.dirMu.Lock()
+		defer d.dirMu.Unlock()
+		for _, c := range d.childMap {
+			c.DecRef(ctx)
+		}
+		d.childMap = nil
+	})
```

Comment:
Why does it appear that the parent has taken a ref on the child?

The cleanup here that needs to happen is that `d` needs to delete itself from `d.parent.childMap`.

If the above comments are addressed, then when `d` is deleted, its d.childMap should be empty.

Reply:
Yes, currently parent will take a ref on the child (as parent has a pointer to the child in its childMap).

As the filesystem is readonly and we will never drop the dentry cache until the filesystem is unmounted (i.e. dentry's refcount will never reach 0 until umount), currently the reference model is pretty simple and works like this:

- The initial refcount of each dentry is 1 which is a reference helped by the parent (so when the refcount is 1, it also means that this dentry is a cached dentry, i.e. not in use).

- When a dentry is used (e.g. opened by someone), its refcount will be increased and the new reference is held by the caller.

- Root dentry's refcount is 2. One reference is returned to the caller of `GetFilesystem()`, and the other one is held by `fs`.

So the above cleanup will happen only when the filesystem is unmounted, which will release all the cached dentries of this filesystem. And in this cleanup procedure, parent's refcount reaches zero first which drives it to drop the references of its children.

Reply:
I see. That can lead to unbounded memory growth in the sentry (due to ever growing dentry tree). I think we can keep this reference model for now (initial version), but ideally, we should have a dentry LRU cache, similar to what fsimpl/gofer does. It caches 1000 *leaf* dentry nodes (or dentries with 0 references). Linux also solves this memory usage problem with dentry cache (code lives in fs/dcache.c). Maybe we can leave context around this in comments, that right now we intend to have an ever growing dentry tree, but work is needed to add a dentry cache.

With VFS2, the filesystem implementation needs to manage dentry logic... it is not handled in VFS layer. Hence will need to duplicate a lot of logic from fsimpl/gofer.

Reply:
Yes, I totally agree. Ideally, we should have a dentry LRU cache. I will leave context around this in comments for now. Thanks!